lib/Sema/CSGen.cpp

```diff
@@ -1937,40 +1937,30 @@ namespace {
 
     Type visitUnresolvedSpecializeExpr(UnresolvedSpecializeExpr *expr) {
       auto baseTy = CS.getType(expr->getSubExpr());
+      Type specializeTy = nullptr;
 
-      if (baseTy->isTypeVariableOrMember()) {
-        return baseTy;
-      }
-
-      // We currently only support explicit specialization of generic types.
-      // FIXME: We could support explicit function specialization.
-      auto &de = CS.getASTContext().Diags;
-      if (baseTy->is<AnyFunctionType>()) {
+      if (baseTy->isTypeVariableOrMember() || baseTy->is<AnyFunctionType>()) {
+        specializeTy = baseTy;
+      } else if (AnyMetatypeType *meta = baseTy->getAs<AnyMetatypeType>()) {
+        specializeTy = meta->getInstanceType();
+      } else {
+        // FIXME: If the base type is a type variable, constrain it to a
+        // metatype of a bound generic type.
+        auto &de = CS.getASTContext().Diags;
         de.diagnose(expr->getSubExpr()->getLoc(),
-                    diag::cannot_explicitly_specialize_generic_function);
+                    diag::not_a_generic_definition);
         de.diagnose(expr->getLAngleLoc(),
                     diag::while_parsing_as_left_angle_bracket);
         return Type();
```

Comment:
If you want, while here it would be nice to get rid of this diagnostic logic; CSGen should never emit diagnostics since it can be run multiple times for conjunction elements. Moreover, CSGen should never fail since it prevents a valid solution from being formed, meaning we emit an unhelpful "unable to infer" diagnostic for e.g:

```swift
func foo(_ x: Int) {
  _ = { // error: unable to infer closure type without a type annotation
    _ = x<String> // error: cannot specialize a non-generic definition
  }
}
```

Instead we should record a fix, maybe we can rely on the diagnostic logic in the specialization constraint for that.

Reply:
I can look at that again -- I tried removing this logic and it leads to some regressions (quoted from above: "I left the check in CSGen for known non-function types because diagnoses pointing at the type-parameter bracket is better for confusing situations like `Parse/generic_disambiguation.swift:29`").  But maybe work-around-able.

Reply:
Yeah, `addSpecializationConstraint` immediately tries to resolve the typereprs in the list, so in the sort of case we are talking about, if you tried to add the constraint you'd get a `cannot find type 'b' in scope` diagnosis immediately from CSGen.

Reply:
Hmm I see, you could always record the fix directly in CSGen to bypass that

Reply:
Actually, the fallout is not bad from that. So long as you get the note telling you the compiler parsed this as a specialization, the can't-find-type error is fine, and this does improve your example.

Added this as a second commit so you can see the differences, will squash if/when approved.

Reply:
(This fixes the cases where there are conjunctions involved in trying to specialize a non-generic type. It does NOT fix when there are conjunctions involved in trying to specialize with a type in the list that doesn't exist).